pkg/cmd/cli/cmd/startbuild.go

```diff
@@ -479,31 +479,56 @@ func streamPathToBuild(git git.Repository, in io.Reader, out io.Writer, client o
 				if gitErr != nil {
 					return nil, fmt.Errorf("the directory %q is not a valid Git repository: %v", clean, gitErr)
 				}
-				pr, pw := io.Pipe()
-				go func() {
-					if err := git.Archive(clean, options.Commit, "tar.gz", pw); err != nil {
-						pw.CloseWithError(fmt.Errorf("unable to create Git archive of %q for build: %v", clean, err))
-					} else {
-						pw.CloseWithError(io.EOF)
-					}
-				}()
-				r = pr
+
+				// Create a temp directory to move the repo contents to
+				tempDirectory, err := ioutil.TempDir(os.TempDir(), "oc_cloning_"+options.Commit)
+				if err != nil {
+					return nil, err
+				}
+				defer os.RemoveAll(tempDirectory)
+
+				// We only want to grab the contents of the specified commit, with
+				// submodules included
+				cloneOptions := git.CloneOptions{
+					Recursive: true,
+					Shallow:   false,
+					Quiet:     true,
+				}
+
+				if verbose := glog.V(3); !verbose {
+					cloneOptions.Quiet = false
+				}
+
+				// Clone the repository to a temp directory for future tar-ing
+				if err := repo.CloneWithOptions(tempDirectory, clean, cloneOptions); err != nil {
+					return nil, err
+				}
+				if err := repo.Checkout(tempDirectory, commit); err != nil {
+					return nil, err
+				}
+				if err := os.RemoveAll(filepath.Join(tempDirectory, ".git")); err != nil {
+					return nil, err
+				}
+
+				// We'll continue to use tar on the cloned directory
+				clean = tempDirectory
 
 			} else {
 				fmt.Fprintf(out, "Uploading directory %q as binary input for the build ...\n", clean)
-
-				pr, pw := io.Pipe()
-				go func() {
-					w := gzip.NewWriter(pw)
-					if err := tar.New().CreateTarStream(path, false, w); err != nil {
-						pw.CloseWithError(err)
-					} else {
-						w.Close()
-						pw.CloseWithError(io.EOF)
-					}
-				}()
-				r = pr
 			}
+
+			pr, pw := io.Pipe()
+			go func() {
+				w := gzip.NewWriter(pw)
+				if err := tar.New().CreateTarStream(path, false, w); err != nil {
```

Comment:
(which is why the code is not streaming an empty dir like i thought it was, it's streaming the wrong dir instead)